src/Microsoft.OData.Core/UriParser/Resolver/CaseInsensitiveSchemaElementsCache.cs

```diff
@@ -0,0 +1,136 @@
+            {
+                return results;
+            }
+
+            return emptyList;
+        }
+
+        /// <summary>
+        /// Find all schema elements of type <typeparamref name="T"/> that match the specified <paramref name="qualifiedName"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of schema elements to match.</typeparam>
+        /// <param name="qualifiedName">The case-insensitive fully qualified name to match.</param>
+        /// <returns>The list of elements of type <typeparamref name="T"/> that matched the <paramref name="qualifiedName"/>.</returns>
+        public IReadOnlyList<T> FindElementsOfType<T>(string qualifiedName) where T : IEdmSchemaElement
+        {
+            IReadOnlyList<IEdmSchemaElement> elements = FindElements(qualifiedName);
+
+            List<T> results = new List<T>();
+            for (int i = 0; i < elements.Count; i++)
+            {
+                if (elements[i] is T element)
+                {
+                    results.Add(element);
+                }
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Find a unique element of type <typeparamref name="T"/> that matches the <paramref name="qualifiedName"/>.
+        /// And exception is thrown if duplicates are found.
+        /// </summary>
+        /// <typeparam name="T">The type of element to match.</typeparam>
+        /// <param name="qualifiedName">The case-insensitive fully qualified name to match.</param>
+        /// <param name="duplicateErrorFunc">A function that generates an error message if a duplicate is found.</param>
+        /// <returns>The element that was found, or `null` if no match was found.</returns>
+        /// <exception cref="ODataException">Thrown if duplicate matches were found.</exception>
+        public T FindSingleOfType<T>(string qualifiedName, Func<object, string> duplicateErrorFunc) where T : IEdmSchemaElement
+        {
+            IReadOnlyList<IEdmSchemaElement> elements = FindElements(qualifiedName);
+
+            if (elements.Count == 0)
+            {
+                return default;
+            }
+
+            T match = default;
+            for (int i = 0; i < elements.Count; i++)
+            {
+                if (elements[i] is T element)
+                {
+                    if (match == null)
+                    {
+                        match = element;
+                    }
+                    else
+                    {
+                        throw new ODataException(duplicateErrorFunc(qualifiedName));
+                    }
+                }
+            }
+
+            return match;
+        }
+        
+        private static void PopulateSchemaElements(IEdmModel model, Dictionary<string, List<IEdmSchemaElement>> cache)
+        {
+            foreach (IEdmSchemaElement element in model.SchemaElements)
+            {
+                string normalizedKey = element.FullName().ToUpperInvariant();

```

Comment:
Remove the `ToUpperInvariant` allocations: the cache dictionary now uses `OrdinalIgnoreCase` comparison so it's not needed.

Reply:
Thanks for the catch. Must have forgotten this.